src/boot/efi/chid.c

```diff
@@ -0,0 +1,385 @@
+                                info->family);
+                break;
+        case 13:
+                status = hash_strings_sha1(
+                                &hash,
+                                (uint8_t *) &namespace,
+                                sizeof(namespace),
+                                5,
+                                info->manufacturer,
+                                L"&",
+                                info->baseboard_manufacturer,
+                                L"&",
+                                info->baseboard_product);
+                break;
+        case 14:
+                status = hash_strings_sha1(
+                                &hash, (uint8_t *) &namespace, sizeof(namespace), 1, info->manufacturer);
+                break;
+        default:
+                return EFI_SUCCESS; /* Just keep empty to prevent match. */
+        }
+        if (EFI_STATUS_IS_ERROR(status))
+                return status;
+
+        memcpy(chid, hash, sizeof(*chid));
+
+        /* Convert the resulting CHID back to little-endian: */
+        chid->Data1 = __builtin_bswap32(chid->Data1);
+        chid->Data2 = __builtin_bswap16(chid->Data2);
+        chid->Data3 = __builtin_bswap16(chid->Data3);
+
+        /* set specific bits according to RFC4122 Section 4.1.3 */
+        chid->Data3 = (chid->Data3 & 0x0fff) | (5 << 12);
+        chid->Data4[0] = (chid->Data4[0] & 0x3f) | 0x80;
+
+        return EFI_SUCCESS;
+}
+
+/**
+ * populate_board_hwids() - Read board SMBIOS and produce an array of CHID values.
+ * @hwids:  Pointer to an array of 12 chids to be filled.
+ */
+static EFI_STATUS populate_board_hwids(EFI_GUID *hwids) {
+        EFI_STATUS status;
+        _cleanup_(smbios_info_clear) SmbiosInfo info = {};
+        int i;
+
+        if (!hwids)
+                return EFI_INVALID_PARAMETER;
+
+        status = populate_smbios_info(&info);
+        if (EFI_STATUS_IS_ERROR(status))
+                return status;
+
+        for (i = 0; i < 15; ++i) {
+                status = get_chid(&info, i, &hwids[i]);
+                if (EFI_STATUS_IS_ERROR(status))
+                        goto exit;
+        }
+
+exit:
+        return status;
+}
+
+typedef struct Device {
+        const char16_t name[128];
+        const char compatible[128];
+        EFI_GUID hwids[32];
+} Device;
+
+EFI_STATUS hwid_match(const void *hwids_buffer, size_t hwids_length, const char **compatible) {
+        EFI_STATUS status;
+
+        const Device *devices = hwids_buffer;
+        size_t n_devices = hwids_length / sizeof(*devices);
+
+        assert(hwids_length % sizeof(*devices) == 0);
+        assert(n_devices > 0);
+        assert(devices);
+
+        static const Device *cached_devices = NULL;
+        static size_t cached_index = 0;
+
+        EFI_GUID hwids[15] = { 0 };
+        int priority[] = { 3, 6, 8, 10, 4, 5, 7, 9, 11 }; /* From most to least specific. */
+
+        if (cached_devices == devices) {
```

Comment:
hmm this kind of caching, does that really make sense? i mean, afaics there's only one callsite for the function, hence might as well just cache locally there?

Reply:
Well, yeah, but having it here make working with this function a bit easier since dtb might not even present on some platforms.